src/Build/Evaluation/ProjectRootElementCache.cs

```diff
@@ -158,91 +209,91 @@ internal ProjectRootElementCache(bool autoReloadFromDisk, bool loadProjectsReadO
+                        return projectRootElement;
                     }
                 }
+            }
 
-                if (projectRootElement == null && openProjectRootElement != null)
-                {
-                    projectRootElement = openProjectRootElement(projectFile, this);
+            // Use openProjectRootElement to reload the element if the cache element does not exist or need to be reloaded.
+            if (projectRootElement == null || projectRootElementIsInvalid)
+            {
+                // We do not lock loading with common _locker of the cache, to avoid lock contention.
+                // Decided also not to lock this section with the key specific locker to avoid the overhead and code overcomplication, as
+                // it is not likely that two threads would use Get function for the same project simulteniously and it is not a big deal if in some cases we load the same project twice.
 
-                    ErrorUtilities.VerifyThrowInternalNull(projectRootElement, "projectRootElement");
-                    ErrorUtilities.VerifyThrow(projectRootElement.FullPath == projectFile, "Got project back with incorrect path");
-                    ErrorUtilities.VerifyThrow(_weakCache.Contains(projectFile), "Open should have renamed into cache and boosted");
-                }
-                else if (projectRootElement != null)
-                {
-                    DebugTraceCache("Satisfied from XML cache: ", projectFile);
-                    BoostEntryInStrongCache(projectRootElement);
-                }
+                ProjectRootElement newProjectRootElement = openProjectRootElement(projectFile, this);
+                ErrorUtilities.VerifyThrowInternalNull(newProjectRootElement, "projectRootElement");
+                ErrorUtilities.VerifyThrow(newProjectRootElement.FullPath == projectFile, "Got project back with incorrect path");
 
                 // An implicit load will never reset the explicit flag.
-                if (projectRootElement != null && isExplicitlyLoaded)
+                if (isExplicitlyLoaded)
                 {
-                    projectRootElement.MarkAsExplicitlyLoaded();
+                    newProjectRootElement?.MarkAsExplicitlyLoaded();
                 }
 
-                return projectRootElement;
+                lock (_locker)
+                {
+                    // Update cache element.
+                    // It is unlikely, but it might be that while without the lock, the projectRootElement in cache was updated by another thread.
+                    // And here its entry will be replaced with newProjectRootElement. This is fine:
+                    // if newProjectRootElement is out of date (so, it changed since the time we loaded it), it will be updated the next time some thread calls Get function.
+                    AddEntry(newProjectRootElement);
+                    projectRootElement = newProjectRootElement;
+                }
```

Comment:
nit: `projectRootElement = newProjectRootElement` doesn't have to run under the lock and `AddEntry` takes the lock internally, so this can be simplified to only call `AddEntry`.